tests/virt_autotest/esxi_open_vm_tools.pm

```diff
@@ -0,0 +1,217 @@
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Run open-vm-tools testing against VMware ESXi
+# Maintainer: Nan Zhang <nan.zhang@suse.com>
+
+use base 'consoletest';
+use strict;
+use warnings;
+use testapi;
+use utils;
+use virt_autotest::esxi_utils;
+use Time::Local;
+
+our $VM_POWERED_ON  = 'Powered on';
+our $VM_POWERED_OFF = 'Powered off';
+
+sub run {
+    my $vm_name = console('svirt')->name;
+    my $vm_id   = esxi_vm_get_vmid($vm_name);
+    my $vm_ip   = esxi_vm_public_ip($vm_id);
+    chomp($vm_ip);
+
+    record_info('Guest Name', $vm_name);
+    record_info('Guest ID',   $vm_id);
+    record_info('Guest IP',   $vm_ip);
+
+    die "The variable \$vm_id or \$vm_ip cannot be empty." if ($vm_id eq "" || $vm_ip eq "");
+
+    assert_script_run('zypper -n in open-vm-tools') if (script_run('rpmquery open-vm-tools'));
+    assert_script_run("/usr/bin/vmware-checkvm | grep 'good'");
+    systemctl('status vmtoolsd');
+    systemctl('status vgauthd');
+    systemctl('restart vmtoolsd');
+    assert_script_run("systemctl status vmtoolsd | grep 'Started open-vm-tools'");
+    assert_script_run("/usr/bin/vmtoolsd -v | grep 'VMware Tools daemon, version'");
+    assert_script_run('vmware-toolbox-cmd logging level set vmtoolsd debug');
+    assert_script_run("vmware-toolbox-cmd logging level get vmtoolsd | grep 'vmtoolsd.level = debug'");
+    assert_script_run('vmware-toolbox-cmd logging level set vmtoolsd message');
+    assert_script_run("vmware-toolbox-cmd logging level get vmtoolsd | grep 'vmtoolsd.level = message'");
+
+    do_power_management_tests($vm_id, $vm_ip);
+    do_networking_tests($vm_id, $vm_ip);
+    do_clock_sync_tests($vm_name, $vm_id, $vm_ip);
+}
+
+sub check_vm_power_state {
+    my ($vm_id, $vm_ip, $power_action, $if_pingable, $check_vm_state, $if_vm_state, $expect_vm_state) = @_;
+
+    if (esxi_vm_power_getstate($vm_id) =~ m/$if_vm_state/) {
+        esxi_vm_power_cycle($vm_id, $power_action);
+        wait_for_vm_bootup($vm_ip, $if_pingable);
+    } elsif ($check_vm_state) {
+        die "Not the correct VM state for guest $power_action.";
+    }
+    die "Failed to $power_action the guest." if (!esxi_vm_power_getstate($vm_id) =~ m/$expect_vm_state/);
+}
+
+sub wait_for_vm_bootup {
+    my $vm_ip       = shift;         # vm ip address
+    my $if_pingable = shift;         # 1 or 0
+    my $times       = shift // 6;    # the number of ping execution
+    my $count       = shift // 3;    # ping waits for the count
+    my $interval    = shift // 5;    # wait secs between each ping
+    my $cmd;
+
+    if ($if_pingable) {
+        $cmd = "ping -c$count -i$interval $vm_ip";
+    } else {
+        $cmd = "! ping -c$count -i$interval $vm_ip";
+    }
+
+    while ($times && console('svirt')->run_cmd($cmd)) {
```

Comment:
The function name is wait_for_vm_bootup, I guess you want to wait until vm bootup, that is until ping finally works? When $if_pingable is 0, the cmd is statically "! ping", how can you tell whether vm is up or not? BTW this function can not differ timeout and cmd works, which I guess it should.

Reply:
This function provides two options that depends on what the power action did. For example, to test the VM shutdown,  we need to give 0 for $if_pingable, as we expected the ping command failed to exit the while loop. And also this ping command can be used for waiting VM bootup or shutdown, it will be run at most 90 seconds by default till it's timeout or break. After this period of time, invoking esxi_vm_power_getstate() to check if it's the expected VM power state. I know there are many ways to check VM is up or not, but just want to choose a simple way in this case because most of tests are based on running ssh command. So we could run those checks once the VM network bringing up or down. Maybe we can extend the check points in future.

Reply:
I see, then what this function does actually differs a little from its function name -- wait_for_vm_bootup, right? It is more like wait_for_vm_bootup_or_powerdown?

Reply:
OK, can we name it wait_for_vm_network? Maybe it's better.

Reply:
I do not quite get what the function does just from this name without knowing your previous explanation. Maybe you can also add some comment to tell what the function does?

Reply:
Added comments in this function.